src/Datadog.Trace.ClrProfiler.Managed/Integrations/Testing/NUnitIntegration.cs

```diff
@@ -0,0 +1,358 @@
+            {
+                scope.Span.SetTag(TestTags.Status, TestTags.StatusPass);
+            }
+        }
+
+        /// <summary>
+        /// Wrap the original NUnit.Framework.Internal.Execution.WorkShift.ShutDown method by adding instrumentation code around it
+        /// </summary>
+        /// <param name="workShift">The workshift instance</param>
+        /// <param name="opCode">The OpCode used in the original method call.</param>
+        /// <param name="mdToken">The mdToken of the original method call.</param>
+        /// <param name="moduleVersionPtr">A pointer to the module version GUID.</param>
+        [InterceptMethod(
+            TargetAssembly = NUnitAssembly,
+            TargetType = NUnitWorkShiftType,
+            TargetMethod = NUnitShutdownMethod,
+            TargetMinimumVersion = Major3Minor0,
+            TargetMaximumVersion = Major3,
+            TargetSignatureTypes = new[] { ClrNames.Void })]
+        public static void WorkShift_ShutDown(
+            object workShift,
+            int opCode,
+            int mdToken,
+            long moduleVersionPtr)
+        {
+            if (workShift == null) { throw new ArgumentNullException(nameof(workShift)); }
+
+            Type workShiftType = workShift.GetType();
+            Action<object> execute;
+
+            try
+            {
+                execute = MethodBuilder<Action<object>>
+                    .Start(moduleVersionPtr, mdToken, opCode, NUnitShutdownMethod)
+                    .WithConcreteType(workShiftType)
+                    .WithNamespaceAndNameFilters(ClrNames.Void)
+                    .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorRetrievingMethod(
+                    exception: ex,
+                    moduleVersionPointer: moduleVersionPtr,
+                    mdToken: mdToken,
+                    opCode: opCode,
+                    instrumentedType: NUnitWorkShiftType,
+                    methodName: NUnitShutdownMethod,
+                    instanceType: workShiftType.AssemblyQualifiedName);
+                throw;
+            }
+
+            execute(workShift);
+            SynchronizationContext context = SynchronizationContext.Current;
+            try
+            {
+                // We have to ensure the flush of the buffer after we finish the tests of an assembly.
+                // For some reason, sometimes when all test are finished none of the callbacks to handling the tracer disposal is triggered.
+                // So the last spans in buffer aren't send to the agent.
+                // Other times we reach the 500 items of the buffer in a sec and the tracer start to drop spans.
+                // In a test scenario we must keep all spans.
+                Tracer.Instance.FlushAsync().GetAwaiter().GetResult();
+            }
+            finally
+            {
+                SynchronizationContext.SetSynchronizationContext(context);
```

Comment:
For my learning, how come we need to re-set the synchronization context here? We're waiting on the FlushAsync method synchronously so I assumed we wouldn't have any change in the synchronization context?

Reply:
Probably we don't need it, and in fact I'm missing one line inside the try block, so thank you!

I usually use this pattern when doing sync over async to reduce the possibilities of deadlocks.
```csharp
            SynchronizationContext context = SynchronizationContext.Current;
            try
            {
                SynchronizationContext.SetSynchronizationContext(null);

                [ASYNC CALL].GetAwaiter().GetResult();
            }
            finally
            {
                SynchronizationContext.SetSynchronizationContext(context);
            }
```

Reply:
fixed in: https://github.com/DataDog/dd-trace-dotnet/pull/870/commits/af50684975ef32b3416e4f8567fc0eb48ec5e841